conbench/entities/distribution.py

```diff
@@ -108,109 +111,160 @@ def get_distribution(benchmark_result, limit):
     )
 
 
-def update_distribution(benchmark_result, limit):
-    from ..db import engine
-
-    if benchmark_result.run.commit.timestamp is None:
-        return
-
-    distribution = get_distribution(benchmark_result, limit).first()
+def update_distribution(
+    benchmark_result: "BenchmarkResult", commit_limit: int
+) -> Optional[dict]:
+    """Try to upsert a Distribution table row, with the given benchmark_result as the
+    most recent BenchmarkResult in the distribution.
+
+    Returns a dict of values if those values were upserted, else returns None. Should
+    not raise anything.
+    """
+    try:
+        distribution_query = _distribution_query(benchmark_result, commit_limit)
+    except CantFindAncestorCommitsError as e:
+        log.debug(
+            f"Not updating distribution: couldn't find ancestor commits, because {e}"
+        )
+        return None
 
+    distribution = distribution_query.first()
     if not distribution:
-        return
+        log.debug("Not updating distribution: the distribution query returned 0 rows")
+        return None
 
     values = dict(distribution)
     hardware_hash = values.pop("hash")
     values["hardware_hash"] = hardware_hash
-    values["limit"] = limit
-
-    with engine.connect() as conn:
-        conn.execute(
-            insert(Distribution.__table__)
-            .values(values)
-            .on_conflict_do_update(
-                index_elements=["case_id", "context_id", "commit_id", "hardware_hash"],
-                set_=values,
-            )
+    values["limit"] = commit_limit
+
+    Session.execute(
+        insert(Distribution.__table__)
+        .values(values)
+        .on_conflict_do_update(
+            index_elements=["case_id", "context_id", "commit_id", "hardware_hash"],
+            set_=values,
         )
-        conn.commit()
+    )
+    Session.commit()
+    return values
 
 
-def get_closest_parent(run):
-    commit = run.commit
-    if commit.timestamp is None:
-        return None
+def get_closest_ancestor(
+    benchmark_result: "BenchmarkResult", branch: Optional[str] = None
+) -> Optional[Commit]:
+    """Given a BenchmarkResult, return the most recent ancestor commit on the given
+    branch that also has a BenchmarkResult in the database with the same
+    hardware/case/context. If branch is not given (default), search all branches.
 
-    hardware_entities = Hardware.all(hash=run.hardware.hash)
-    hardware_ids = set([m.id for m in hardware_entities])
+    Return None if one isn't found. Should not raise anything.
+    """
+    from ..entities.benchmark_result import BenchmarkResult
```

Comment:
There would be a circular import due to

https://github.com/conbench/conbench/blob/614e947a1b95525ba48c57ded6559ca8f9de8b82/conbench/entities/benchmark_result.py#L22

I tried reversing the direction of this functional import, but it caused a lot of issues with SQLAlchemy (I think it could be magically using import order to determine how to delete tables?) so I left the direction as it was before this PR. I decided it wasn't worth the time to investigate.